Code/src/main/java/nl/utwente/group10/haskell/type/TypeVar.java

```diff
@@ -207,35 +189,36 @@ public final void setConcreteInstance(ConcreteType type) {
      * @param the other type variable.
      */
     public final void unifyWith(TypeVar other) {
-        this.instance.unifyInstances(other.instance);
+        if (this.instance.internal) {
+            this.instance.unifyWith(other.instance);
+        } else {
+            other.instance.unifyWith(this.instance);
+        }
     }
 
     /**
-     * @return Whether this type variable is constrained.
+     * @return The set of type class constraints associated with this type variable
      */
-    public final boolean hasConstraints() {
-        return !this.instance.constraints.isEmpty();
+    ConstraintSet getConstraints() {
+        return this.instance.constraints;
```

Comment:
How about this method? Can the caller of this method introduce extra constraints on the result it gets?